fix: JSON serialization of `metrics::Examplar` and `trace::span::Link…

demurgos opened the pull request:
* [x] add more tests

Fixes #2065
Design discussion issue (if applicable): N/A

## Changes

This commit fixes open-telemetry/opentelemetry-rust#2065: it uses string serialization for all `TraceId` and `SpanId` fields. In particular, this fixes serialization of `metrics::Examplar` and `trace::span::Link`.

As part of the change, JSON deserialization tests were greatly expanded to test both serialization and deserialization, with support for alternative serializations. The tests are extended to cover all examples from the `opentelemetry-proto` repository, it also adds a test where all values are used (without any default primitives or empty array).

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

TommyCpp commented on opentelemetry-proto/CHANGELOG.md:
```diff
@@ -1,7 +1,9 @@
--  Update protobuf definitions to v1.3.2 [#1945](https://github.com/open-telemetry/opentelemetry-rust/pull/1945)
+
+- Update protobuf definitions to v1.3.2 [#1945](https://github.com/open-telemetry/opentelemetry-rust/pull/1945)
+- Fix JSON serialization of `metrics::Exemplar` and `trace::span::Link`
```
nit: could you link the PR here

demurgos replied:
Fixed: I added the PR link.

lalitb commented:
Thanks @demurgos, this is quite extensive. And also good to have the validation for examples from otel-proto. Couple of nit comments, apart from the CI lint error to be taken care of.

demurgos commented:
Thank you for the feedback, I'll fix the typos and lint errors.

demurgos commented on opentelemetry-proto/tests/grpc_build.rs:
```diff
@@ -41,8 +41,8 @@ fn build_tonic() {
 
-    // optional numeric and String field need to default it to 0 otherwise JSON files without those field
-    // cannot deserialize
+    // optional numeric, string and array fields need to default it their default value otherwise
```
Fixed

demurgos commented on opentelemetry-proto/tests/json_serde.rs:
```diff
@@ -0,0 +1,1482 @@
+      }
+    }
+
+    // request were all fields are set
```
Fixed

demurgos commented:
I fixed the typos, changelog and formatting.

I also fixed the linter error. Some of them were due to feature combinations so I had to add some extra `cfg(...)` in the tests. I also update the linter script to check for the relevant feature combinations (for `logs` and `metrics`).

lalitb commented:
> The remaining error are spurious diffs in the generated protobuf code. I'm not sure if it's just a bit flaky or related to my changes.

This is fixed in #2077. CI should hopefully pass once the conflict is resolved.

demurgos commented:
@lalitb I rebased my PR on top of main. CI is now green.
